test: add automated testing using plenary

compactcode opened the pull request:
## Why

1. To help new contributors understand how the plugin works and what input/output functions expect.
2. To give instant feedback when making changes.

## What

I added some busted style tests according the to the [plenary guide](https://github.com/nvim-lua/plenary.nvim#plenarytest_harness) and run them using the [neotest plenary](https://github.com/nvim-neotest/neotest-plenary) adapter.

![1655336801](https://user-images.githubusercontent.com/125175/173960757-0d6c5a02-afa4-411f-8eb5-3cd6febc4841.png)

I added two test files:

- `init_spec.lua` this would contain high level tests for the public functions that neotest depends on. Ideally stuff like given a test file (e.g. `spec/basic_example_spec.rb`) I expect `discover_positions` to return a tree containing the ids we want to see in the summary window.
- `utils_spec.lua` this would contain unit tests for internal functions that might require a lot of tests that are a pain to test at higher levels.

## Discussion

I'm probably doing some things wrong here so not expecting this to be merged as is. Wanted to get some early feedback on whether this is an idea you be interested in pursuing.

If so I'm happy to flesh the examples out and incorporate any feedback before merging.

compactcode commented on lua/spec/neotest-rspec/init_spec.lua:
```diff
@@ -0,0 +1,17 @@
+local plugin = require('neotest-rspec')
```
I'm not yet sure if this is actually loading the module from inside this project or the one installed by my plugin manager.

olimorris replied:
I think we'll need to clone `neotest` in order to test properly. I clone plenary in my [minimal.vim](https://github.com/olimorris/persisted.nvim/blob/main/tests/minimal.vim) file for my session plugin...but something to worry about later.

compactcode commented on lua/spec/neotest-rspec/init_spec.lua:
```diff
@@ -0,0 +1,17 @@
+
+describe('discover_positions', function()
+  it('matches rspec files', function()
+    assert.equals('something', plugin.discover_positions('./spec/nested_tests/nested_spec.rb'))
```
Currently fails with the following error:

```
neotest/lua/neotest/lib/treesitter/init.lua:138: attempt to yield across C-call boundary
```

Seems to be caused when neotest attempts to sleep:

https://github.com/nvim-neotest/neotest/blob/f26979786377704ce85fe43d453418249b1fe7e4/lua/neotest/lib/treesitter/init.lua#L137-L141

Maybe this could be mocked out.

olimorris replied:
This was the biggie I hit...I wonder if we need to utilise `plenary.async.tests`. I did it recently for my [session plugin](https://github.com/olimorris/persisted.nvim/blob/main/tests/autoload/autoload_allowed_dir_spec.lua). So you call `async.describe` instead. Let me know if that does anything.

compactcode replied:
The async helpers works and I was able to get access to the neotest `Tree` table.

olimorris commented:
Firstly thanks very much for you contribution. I agree with all of this in principle and it was something I started playing around with at the weekend and came across the same errors as you did. I raised a [discussion](https://github.com/nvim-neotest/neotest/discussions/20#discussioncomment-2944890) over in the `neotest` repo too and we could always ask [rcarriga](https://github.com/rcarriga) for his help.

olimorris commented on lua/neotest-rspec/init.lua:
```diff
@@ -23,23 +24,7 @@ end
-    :gsub('<NS>', '')
-    :gsub('<NS>', '')
-    :gsub('</NS>', '')
+  return utils.form_treesitter_id(position_id)
```
I agree with abstracting to a seperate lua file. We could then just remove the local function alltogether and call `utils.form_treesitter_id` in the `NeotestAdapter.discover_positions` method.

Also think we should move `parse_json_output` to the utils file too.

compactcode replied:
Good point that the local doesn't add much value anymore.

I was also thinking that extracting `parse_json_output` would be useful.

compactcode reviewed (commented):
Ok great, I'll proceed and flesh things out a little more :smile:

compactcode commented on lua/spec/neotest-rspec/init_spec.lua:
```diff
@@ -0,0 +1,22 @@
+    local positions = plugin.discover_positions('./spec/nested/basic_spec.rb')
+
+    assert.equals('basic_spec.rb', positions._data.name)
+    assert.equals('Maths', positions._nodes.Maths._data.id)
```
Not quite sure which bits of the Tree structure we care about yet. Also looking for a good way to make assertions on table data in lua.

olimorris replied:
You may need `async.describe` in there too

compactcode commented:
Didn't get much time today and tomorrow will be busy so will try to pick this up again over the weekend.

olimorris commented:
As a heads up, there's likely to be some merge conflicts as I've pushed some changes through in the last 24 hours. Nothing major, just some minor tweaks to formatting and outputing. Let me know if you need me to resolve the conflicts

compactcode commented:
> As a heads up, there's likely to be some merge conflicts as I've pushed some changes through in the last 24 hours. Nothing major, just some minor tweaks to formatting and outputing. Let me know if you need me to resolve the conflicts

Thanks for the notification, the conflicts were straightforward to resolve.

I'm happy enough with the `discover_positions` now. I've also moved `parse_json_output` to utils and added a test for that.

Let me know what you think.

compactcode commented on lua/spec/neotest-rspec/init_spec.lua:
```diff
@@ -0,0 +1,44 @@
+          }
+        }
+      }
+    }
```
I think this gives a nice overview of the important things that this function is supposed to return.

compactcode commented on lua/spec/neotest-rspec/init_spec.lua:
```diff
@@ -0,0 +1,44 @@
+    assert.equals(expected_output[2][1].name, positions[2][1].name)
+    assert.equals(expected_output[2][1].type, positions[2][1].type)
+    assert.equals(expected_output[2][2][1].name, positions[2][2][1].name)
+    assert.equals(expected_output[2][2][1].type, positions[2][2][1].type)
```
Would love something like the rspec [include matcher](https://relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/include-matcher) here.

compactcode commented on lua/neotest-rspec/utils.lua:
```diff
@@ -0,0 +1,62 @@
+
+---@param position_id string
+---@return string
+M.form_treesitter_id = function(position_id)
```
I'd like to refactor this method a bit but don't want to make this PR any more complicated. Will proposed something in a subsequent PR.

olimorris commented:
This looks great. I'll pull it down and test tomorrow.

Keen to get GitHub workflows working for this as well so will put that together in another PR.

Thanks so much.

olimorris commented:
Lovely work. Thanks so much for all of your efforts.
